llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -2263,6 +2263,12 @@ static bool isKnownNonNullFromDominatingCondition(const Value *V,
         return true;
     }
 
+    if (const auto *Div = dyn_cast<BinaryOperator>(U);
+        Div && Div->getOpcode() == BinaryOperator::UDiv &&
+        Div->getOperand(1) == V &&
+        isValidAssumeForContext(cast<Instruction>(U), CtxI, DT))
```

Comment:
`isValidAssumeForContext` assumes the first argument is an assume intrinsic. It is incorrect to pass a udiv.

Reply:
Hmm, I am not sure of any other way to check for post-dominance.

Reply:
Maybe `DT->dominates(cast<Instruction>(U), CtxI)`? It guarantees we execute `udiv` before `select`. Then we can simplify the pattern in `simplifySelectWithICmpCond`.

Reply:
In this case, `CtxI` is always the `icmp`, so the `dominates` call will only work when the `icmp` dominates the `udiv` and not the other way round.

Reply:
isValidAssumeForContext() should be fine to use for non-assumes as well. The key property of assumes here is that violating the fact results in immediate undefined behavior, which is also true for divisions.

Though I guess the isEphemeralValueOf() part of the function is only relevant for assumes...